docs/core/knowledge-bases.rst

```diff
@@ -0,0 +1,383 @@
+The action is also able to resolve certain mention of objects, such as ``the first one`` or ``that restaurant``.
+You can find a complete example in ``examples/knowledge_base_bot``.
+
+.. note::
+   There is a tutorial `here <https://blog.rasa.com/integrating-rasa-with-knowledge-bases/>`_ about how to use
+   knowledge bases in custom actions. The tutorial will explain in detail the implementation behind
+   ``ActionQueryKnowledgeBase``.
+
+
+Create a KnowledgeBase
+----------------------
+
+The data you will use to answer the user's request comes from a knowledge base.
+A knowledge base can be used to store complex data structures.
+If you just have some data points that fit into memory, you can use our ``InMemoryKnowledgeBase`` implementation.
+To initialize an ``InMemoryKnowledgeBase`` you need to provide the data in a json file.
+
+Let's take a look at an example:
+
+.. code-block:: json
+
+    {
+        "restaurant": [
+            {
+                "id": 0,
+                "name": "Donath",
+                "cuisine": "Italian",
+                "outside-seating": true,
+                "price-range": "mid-range"
+            },
+            {
+                "id": 1,
+                "name": "Berlin Burrito Company",
+                "cuisine": "Mexican",
+                "outside-seating": false,
+                "price-range": "cheap"
+            },
+            {
+                "id": 2,
+                "name": "I due forni",
+                "cuisine": "Italian",
+                "outside-seating": true,
+                "price-range": "mid-range"
+            }
+        ],
+        "hotel": [
+            {
+                "id": 0,
+                "name": "Hilton",
+                "price-range": "expensive",
+                "breakfast-included": true,
+                "city": "Berlin",
+                "free-wifi": true,
+                "star-rating": 5,
+                "swimming-pool": true
+            },
+            {
+                "id": 1,
+                "name": "Hilton",
+                "price-range": "expensive",
+                "breakfast-included": true,
+                "city": "Frankfurt am Main",
+                "free-wifi": true,
+                "star-rating": 4,
+                "swimming-pool": false
+            },
+            {
+                "id": 2,
+                "name": "B&B",
+                "price-range": "mid-range",
+                "breakfast-included": false,
+                "city": "Berlin",
+                "free-wifi": false,
+                "star-rating": 1,
+                "swimming-pool": false
+            },
+        ]
+    }
+
+The above json file contains data about restaurants and hotels.
+The json structure should contain a key for every object type, i.e. "restaurant" and "hotel".
+Every object type maps to a list of objects.
+
+Once the data are defined in a json file, called, for example, ``data.json``, you can create your
```

Comment:
did we get rid of the `schema` part? if so that's a nice simplification

Reply:
yes, split it up into separate methods and set default values